prusti-viper/src/encoder/mir/contracts/interface.rs

```diff
@@ -0,0 +1,376 @@
+use super::borrows::{BorrowInfo, BorrowInfoCollectingVisitor};
+use crate::encoder::{
+    encoder::SubstMap,
+    errors::{EncodingError, EncodingResult},
+    places, Encoder,
+};
+use log::{debug, trace};
+use prusti_interface::specs::typed;
+use rustc_hash::FxHashMap;
+use rustc_hir::{
+    def_id::{DefId, LocalDefId},
+    Mutability,
+};
+use rustc_middle::{mir, ty, ty::FnSig};
+use std::{cell::RefCell, fmt};
+
+pub(crate) trait ContractsEncoderInterface<'tcx> {
+    /// Get a contract for a procedure's definition site, using MIR types for
+    /// locals and places. Results are cached per DefId.
+    fn get_mir_procedure_contract_for_def(
+        &self,
+        proc_def_id: DefId,
+    ) -> EncodingResult<ProcedureContractMirDef<'tcx>>;
+
+    /// Get a contract for a procedure's definition site, using the
+    /// `encoder::places` abstraction for locals and places. Results are cached
+    /// per DefId.
+    fn get_procedure_contract_for_def(
+        &self,
+        proc_def_id: DefId,
+    ) -> EncodingResult<ProcedureContract<'tcx>>;
+
+    /// Get a contract for a call to a procedure, with particular argument and
+    /// result substitutions.
+    fn get_procedure_contract_for_call(
+        &self,
+        proc_def_id: DefId,
+        args: &[places::Local],
+        target: places::Local,
+        tymap: SubstMap<'tcx>, // TODO: Simplify these params
+        substs: ty::subst::SubstsRef<'tcx>,
+    ) -> EncodingResult<ProcedureContract<'tcx>>;
+}
+
+#[derive(Default)]
+pub(crate) struct ContractsEncoderState<'tcx> {
+    encoded_contracts: RefCell<FxHashMap<DefId, EncodingResult<ProcedureContractMirDef<'tcx>>>>,
+}
+
+/// Contract of a specific procedure. It is a separate struct from a
+/// general procedure info because we want to be able to translate
+/// procedure calls before translating call targets.
+#[derive(Clone, Debug)]
+pub struct ProcedureContractGeneric<L, P>
```

Comment:
It would be nice to leave in this file only the state struct and the interface struct and move all other items to a different file.